File: aoc_2018/day01.py
# find total for one pass
def part1(changes):
    result = 0
    for change in changes:
        result += int(change)
    return result

# find first repeated intermediate result for any number of passes
def part2(changes):
    result = 0
    results = { result }
    while True:
        for change in changes:
            result += int(change)
            if result in results:
                return result
            results.add(result)

File: aoc_2018/kt18/src/main/kotlin/com/github/dstine/aoc/kt18/Day01.kt
package com.github.dstine.aoc.kt18

// added in Kotlin 1.3
import kotlin.sequences.sequence

fun part1(changes: List<String>): Int =
    changes.asSequence()
        .map { it.toInt() }
        .sum()

// https://stackoverflow.com/a/48007581
fun <T> Sequence<T>.repeat() = sequence { while (true) yieldAll(this@repeat) }

// find first repeated intermediate result for any number of passes
fun part2(changes: List<String>): Int {
    var result = 0
    val results = mutableSetOf(result)
    changes.asSequence()
        .map { it.toInt() }
        .repeat()
        .forEach {
            result += it
            if (result in results) {
                return result
            }
            results.add(result)
        }

    throw RuntimeException("should never get here")
}
